wandb/sdk/launch/launch_add.py

```diff
@@ -0,0 +1,99 @@
+import json
+from typing import Any, Dict
+import sys
+
+import wandb
+import wandb.apis.public as public
+from wandb.apis.internal import Api
+from wandb.sdk.launch.utils import construct_launch_spec, set_project_entity_defaults
+
+
+def push_to_queue(api: Api, queue: str, launch_spec: Dict[str, Any]) -> Any:
+    try:
+        res = api.push_to_run_queue(queue, launch_spec)
+    except Exception as e:
+        print("Exception:", e)
+        return None
+    return res
+
+
+def launch_add(
+    uri: str = None,
+    config: Dict[str, Any] = None,
+    project: str = None,
+    entity: str = None,
+    queue: str = None,
+    resource: str = None,
+    entry_point: str = None,
+    name: str = None,
+    version: str = None,
+    docker_image: str = None,
+    params: Dict[str, Any] = None,
+) -> "public.QueuedJob":
+    api = Api()
+    return _launch_add(
+        api,
+        uri,
+        config,
+        project,
+        entity,
+        queue,
+        resource,
+        entry_point,
+        name,
+        version,
+        docker_image,
+        params,
+    )
+
+
+def _launch_add(
+    api: Api,
+    uri: str,
+    config: Dict[str, Any],
+    project: str,
+    entity: str,
+    queue: str,
+    resource: str,
+    entry_point: str,
+    name: str,
+    version: str,
+    docker_image: str,
+    params: Dict[str, Any],
+) -> "public.QueuedJob":
+
+    resource = resource or "local"
+    if config is not None:
+        with open(config, "r") as f:
+            launch_config = json.load(f)
+    else:
+        launch_config = {}
+
+    project, entity, _ = set_project_entity_defaults(
```

Comment:
the args/config thing is handled in `construct_launch_spec` and the rest of the default stuff is handled after the job is pulled by the agent in `create_project_from_spec` -- i wanted a clean separation where everything that goes into the "launch spec" is exactly what the user passes in, and then once the agent pulls it to prep for running we make changes set defaults etc. sorry that wasn't clear i should've put up a comment but this shouldn't be here

Reply:
launch_add requires the launch_spec to have a project/entity populated in the spec to push it, since these values are used in the push_to_queue function. So this is necessary to do here, or we can make them required arguments.